nomad/worker.go

```diff
@@ -414,8 +415,26 @@ func (w *Worker) run() {
 		w.setWorkloadStatus(WorkloadWaitingForRaft)
 		snap, err := w.snapshotMinIndex(waitIndex, raftSyncLimit)
 		if err != nil {
-			w.logger.Error("error waiting for Raft index", "error", err, "index", waitIndex)
-			w.sendNack(eval, token)
+			var timeoutErr ErrMinIndexDeadlineExceeded
+			if errors.As(err, &timeoutErr) {
+				w.logger.Warn("timeout waiting for Raft index required by eval", "index", waitIndex, "timeout", raftSyncLimit)
+				w.sendNack(eval, token)
+
+				// Timing out here means this server is woefully behind the leader's
+				// index. This can happen when a new server is added to a cluster and
+				// must initially sync the cluster state.
+				// Backoff dequeuing another eval until there's some indication this
+				// server would be up to date enough to process it.
+				const slowServerSyncLimit = 10 * raftSyncLimit
+				if _, err := w.snapshotMinIndex(waitIndex, slowServerSyncLimit); err != nil {
+					w.logger.Warn("server is unable to catch up to last eval's index", "error", err)
```

Comment:
If we get this warning, it's going to be quite a while after the previous log line (50s), so it might be hard to correlate with the previous one unless you're already familiar with the behavior, especially because we'll have quite a few workers running concurrently which should all have the same problem. Given this should be rare, maybe we can add more context here around which eval we previously dequeued?

Reply:
Re-reading this I'm realizing we don't care about the previous eval by the time we've hit this, as we've already nacked it and we have the index in the error message.